### DISCUSSION TITLE
Is there any way to give hints to the error recovery process?

### DISCUSSION THREAD

**Discussion by ahelwer** [tree-sitter/tree-sitter#1205] [Q&A] [answered]
Is there any way to identify "anchor" definitions to the error recovery process? For example, in my language I have some constructs that are roughly analogous to new function definitions. They use a nearly-unique token and would make an ideal error recovery point for the rest of the file. How does the error recovery process work, roughly? Does it replay some sequence with GLR parsing and try to maximize the score? So would it work to give a really high dynamic precedence to my "anchor" definition?

**Comment by maxbrunsfeld** [comment]
I'm interested in adding this for my own use cases, and may do it in the near future. It will require a bit of design work and some kinda tricky changes to the parsing code, but I definitely think it will happen.

**Reply by maxbrunsfeld** [reply, marked as answer]
I don't know what the design will be - I'd like to make it as *automatic* as possible (as opposed to being very configurable). This blog post has a good description of the limitations I'd like to address: https://rust-analyzer.github.io/blog/2020/09/16/challeging-LR-parsing.html.

Basically, we currently err on the side of *skipping* tokens and subtrees that can't fit into our grammar. But in many cases, it would be better to *insert* more "missing" tokens in order to complete what the user was in the process of typing.

**Reply by ahelwer** [reply]
@maxbrunsfeld thanks, that was a really interesting blog post, as well as the linked [Which Parsing Approach?](https://tratt.net/laurie/blog/entries/which_parsing_approach.html) one. Do you know of any good texts which cover all these different approaches and their benefits/drawbacks, for someone looking at hand-writing a parser to follow the tree-sitter version?

**Comment by stephe-ada-guru** [comment]
Max Brunsfeld ***@***.***> writes:

> I don't know what the design will be - I'd like to make it as
> *automatic* as possible (as opposed to being very configurable). This
> blog post has a good description of the limitations I'd like to
> address:
> https://rust-analyzer.github.io/blog/2020/09/16/challeging-LR-parsing.html.
>
> Basically, we currently err on the side of *skipping* tokens and
> subtrees that can't fit into our grammar. But in many cases, it would
> be better to *insert* more "missing" tokens in order to complete what
> the user was in the process of typing.

You might be interested in the algorithm I'm using for that (draft of
unpublished paper):

https://stephe-leake.org/ada/error_correction_algorithm.pdf

The parser generator home page is at:
https://stephe-leake.org/ada/wisitoken.html

In particular, the "minimal_complete" algorithm can fill in all the
missing "end"s in statements (ie "end if;", "end loop;" etc in Ada; "}"
in Java).

It works very well with batch parsing, and produces a syntax tree that
is good enough for computing indentation.

As written, it does not handle incremental parsing. I'm working on the
code to implement that; so far, the changes to the error recovery are
simple; in the parsing step, nonterminals can be shifted.

This does complicate Edit_Tree a lot, because of the inserted virtual
tokens; I'm currently finding lots of corner cases while running my test
suite.

-- 
-- Stephe
